### ISSUE TITLE
[FR] Solvate with fixed number of waters

### ISSUE THREAD

**Issue by cxhernandez** [pandegroup/openmm#886]
I think one thing that's been a big headache for people who are trying to make reproducible clones of simulations has been getting a fixed number of solvent molecules. I think It would be nice to add this functionality into `Modeller`. 

I have a prototype of the code on this [gist](https://gist.github.com/cxhernandez/5afae8cce3dcb034ee95), which is based off of @choderalab's [Ensembler](https://github.com/choderalab/ensembler/blob/aa3cfe690017ad0bb3dce49158ce8e6f31fb0127/ensembler/refinement.py#L565-L683) code.

**Comment by kyleabeauchamp** [comment]
Hmm, possible locations for this might be:

OpenMM
OpenMMTools
Parmed
OpenMolTools

The advantage of doing this downstream of OpenMM is that you could possible have access to extra dependencies...

**Comment by rmcgibbo** [comment]
I've also needed to solvate many copies of the same system (different conformations) with the same number of water molecules, and wasted a couple hours writing a script for it. +1 to putting some version of this functionality into the `Modeller` class.

**Comment by rmcgibbo** [comment]
Neither @cxhernandez's code nor the ensembler code uses anything outside of OpenMM, AFAICT.

**Comment by rmcgibbo** [comment]
IMO, the most logical place for this functionality is as a method on the modeller class, which happens to be in OpenMM.

**Comment by kyleabeauchamp** [comment]
+1

**Comment by jchodera** [comment]
See also:
* https://github.com/pandegroup/openmm/issues/312
* https://github.com/pandegroup/openmm/issues/518

**Comment by rmcgibbo** [comment]
Oh right. If there's a consensus that @danielparton and @cxhernandez's approach is the right one, then that probably solves the biggest hurdle. @cxhernandez: do you want to take a stab at an OpenMM PR?

**Comment by cxhernandez** [comment]
Oh wow, I completely forgot about #312

**Comment by cxhernandez** [comment]
@rmcgibbo: yeah sure

**Comment by swails** [comment]
+1 for `Modeller`

(hey wait guys!  i'm here!)

**Comment by cxhernandez** [comment]
also any suggestions on what this kwarg should be named to keep in line with @peastman's conventions?

**Comment by peastman** [comment]
Kyle tried to implement this once before, but gave up on it when it started to get complicated: #451.  It might be simplest to just redefine the problem.  Instead of specifying the maximum number of **water molecules**, specify the maximum number of **molecules** (both water and ions) to add.

**Comment by rmcgibbo** [comment]
@peastman: did you try looking at the solutions at @cxhernandez and @danielparton implemented? I think they already have solved the problem. Maybe not the most elegant, but it looks passable.

**Comment by jchodera** [comment]
I think both of those are based on my same original implementation. Unless
there are differences/improvements I am missing?

**Comment by peastman** [comment]
#451 and #888 are solving slightly different problems.  One tries to evenly distribute a fixed number of waters over a box of specified volume.  The other tries (though not entirely successfully) to densely pack the specified number of waters into a box of unknown volume.

I don't really have a good sense of what the intended use cases for this feature are.  Can we enumerate those?  That will dictate how to approach it.

**Comment by jchodera** [comment]
My use cases:
* I have N initial conformations of the same biomolecule, and want to solvate them in a box with the same number of water molecules so that I can combine data from multiple simulations.  I can't do this rigorously if they differ in the number of molecules.
* I want to make a solvent box containing only solvent that has precisely N molecules of solvent in it so I can study liquid properties.

**Comment by kyleabeauchamp** [comment]
So the second use case is perhaps better solved using packmol and our wrappers for it (https://github.com/choderalab/openmoltools/blob/master/openmoltools/packmol.py#L28).  I only suggest this in an attempt to help limit the scope of the present work.

**Comment by peastman** [comment]
That's not quite what I meant.  *Why* do you want the box to contain precisely N molecules?  Are there cases where you want to specify a maximum number of molecules (as in #518)?  Is it important for the box size to be fixed?  Is it important for the results to be deterministic?

**Comment by swails** [comment]
> Why do you want the box to contain precisely N molecules?

You can't compare extensive quantities between simulations of the same solute when the atom counts are different (e.g., potential energies).

**Comment by peastman** [comment]
Also, is the goal just to call this several times for different conformations and always get the same number of waters and ions (but you don't care exactly what that number is)?  Or is it essential for that number to be precisely what you specify?  Is it better to specify the number of waters, or the number of waters+ions, or does it not matter?

**Comment by peastman** [comment]
> You can't compare extensive quantities between simulations of the same solute when the atom counts are different (e.g., potential energies).

Sorry, I wasn't clear.  I was referring to John's second point, not his first one.

**Comment by jchodera** [comment]
>  Why do you want the box to contain precisely N molecules? Are there cases where you want to specify a maximum number of molecules (as in #518)? Is it important for the box size to be fixed? Is it important for the results to be deterministic?

One of many reasons: Reweighting techniques (e.g. MBAR) require you have a potential energy function U(x;lambda) where you can feed it different configurations `x_n, n=1..N` that all have the same dimension.  Work on transdimensional reweighting is ongoing, but we don't have the theory for that yet.  Even if we did, it would almost certainly be much more complex.

**Comment by jchodera** [comment]
Commented before I could address the other questions!

> Are there cases where you want to specify a maximum number of molecules (as in #518)?

I'm not sure what the use cases for desiring a *maximum* number of waters are.

> Is it important for the box size to be fixed?

If we wanted to run NVT simulations, yes; but the tool [PackMol](http://www.ime.unicamp.br/~martinez/packmol/) which Kyle has wrapped in [`openmoltools`](https://github.com/choderalab/openmoltools) already handles this case.

My primary use cases are all NPT simulations where the box size can be variable.

> Is it important for the results to be deterministic?

Good question.  I don't think there's a strong requirement for that, though I could see some utility in this.

**Comment by jchodera** [comment]
But, as @kyleabeauchamp, he already has a tool for preparing neat liquids in [`openmoltools`](https://github.com/choderalab/openmoltools) if we want to reduce the scope of this.

**Comment by peastman** [comment]
Let's not worry about reducing scope just yet.  First let's figure out what would be ideal.  Then we can reduce the scope if that turns out to be a problem, but it might not.

**Comment by rmcgibbo** [comment]
My view is that the ideal, or core purpose of the proposed feature, is to have a painless method to solvate `N` distinct conformations of the same protein system with an identical number of water molecules and ions, more or less densely packed around the protein.

The use case for this, as John pointed out, is for properly initializing a series of NPT simulations to be run in parallel.

I don't think performance is a particularly pressing concern.

**Comment by peastman** [comment]
Ok, let me try to summarize what I'm hearing:

1. All the use cases seem to be for specifying *fixed* number of waters, not *maximum* number of waters.

2. In many use cases, it doesn't matter what the exact box volume is.  There are some use cases where it does matter, but those are less important since they can already be addressed with another tool.

No one has expressed any opinions on the question of specifying number of waters versus number of waters+ions.  The latter is easier to implement, but the former can still be done if there's a good reason to make it work that way.

**Comment by rmcgibbo** [comment]
1 + 2. Yep, I agree.

> No one has expressed any opinions on the question of specifying number of waters versus number of waters+ions. The latter is easier to implement, but the former can still be done if there's a good reason to make it work that way.

The overarching goal is that when you solvate `N` different protein conformations, all of the resulting systems are members of the same NPT ensemble, so that would include having the same number of waters and the same number of ions.

**Comment by rmcgibbo** [comment]
We can bikeshed about the details, but does anyone have an alternative algorithm/approach other than what @cxhernandez implemented in #888?

**Comment by peastman** [comment]
Right, that's a given.  Calling it with the same arguments, it will always add exactly the same things.  But you don't care whether the number you specified ends up as the number of waters, or the number of waters+ions?

> does anyone have an alternative algorithm/approach other than what @cxhernandez implemented in #888?

Yes, I have a completely different algorithm in mind that will be much more efficient.

**Comment by rmcgibbo** [comment]
> But you don't care whether the number you specified ends up as the number of waters, or the number of waters+ions?

Right. I think either way is fine.

> Yes, I have a completely different algorithm in mind that will be much more efficient.

Oh, cool!

**Comment by cxhernandez** [comment]
> Yes, I have a completely different algorithm in mind that will be much more efficient.

Perfect!

**Comment by peastman** [comment]
Ok, how about this as an API: a new argument called `numAdded` for addSolvent().  It gives the total number of molecules (water and ions) to add.  This would be yet another way to specify the box size.  So you would specify `boxSize` or `boxVectors` or `padding` or `numAdded`, but never two of those.  It would behave exactly like `padding`, except the necessary padding distance would be chosen to be exactly large enough to contain the specified number of molecules.

**Comment by rmcgibbo** [comment]
:+1:

**Comment by jchodera** [comment]
+1 for same number of waters and same number of ions.

**Comment by cxhernandez** [comment]
+1 for `numAdd` flag

**Comment by swails** [comment]
+1

**Comment by jchodera** [comment]
Maybe "totalNumWaters" or something a bit more descriptive than "numAdd"?

**Comment by peastman** [comment]
I intentionally didn't call it that, because it *isn't* the number of waters.

**Comment by jchodera** [comment]
That's fine---just lobbying for a descriptive term, rather than a cryptic argument.

**Comment by peastman** [comment]
Any suggestions for something that is more descriptive but still accurate?

**Comment by cxhernandez** [comment]
`numSolventAdded`?

**Comment by leeping** [comment]
(I've been lurking.) 

I think the user would often like to simulate two different outputs of addSolvent() using the same `System` object.  Thus, the number of solvent + ions needs to be consistent, but the types of molecules / ions added also need to come in a deterministic order.  For example, if we have two conformations of a 100-residue protein and addSolvent() produces a water molecule in residue 101 for the first conformation, then it must also produce a water molecule in residue 101 for the second conformation.

I'm indifferent for the variable name, actually `numAdded` sounds descriptive enough to me.  Along the lines of what Carlos suggested, other possible suggestions are `numMoleculesAdded`, `numToAdd`, `numMoleculesToAdd`, `numSolventToAdd`.

**Comment by jchodera** [comment]
All of @leeping's suggested names are good!  Or if you might want to leave
room for future alternative behaviors, ‘addExactly‘ might be good.

Since crystal waters might be retained, the user might want to instead
specify the total number of solvent molecules to end up with (for
convenience), or easily query how many waters are already there.  I'm
thinking of cases where the input structures may be from different crystal
structures with different numbers of X-ray waters, and if you want to
retain those and end up with the same total number of waters, we need a bit
more than just controlling the number of waters added.

**Comment by peastman** [comment]
Understood.  It will add exactly the same number of waters and ions in exactly the same order.  The only things that differ will be their positions and the box size.

> or easily query how many waters are already there.

    sum(1 for res in topology.residues() if res.name == 'HOH')

**Comment by peastman** [comment]
Fixed by #894.  Does it look ok to everyone?

**Comment by jchodera** [comment]
The `numAdded` flag in #894 looks good---thanks!

Regarding the querying of how much solvent is there already:  While I agree that
```
sum(1 for res in topology.residues() if res.name == 'HOH')
```
is "simple" in a sense (in that it fits on one line), I don't think this qualifies as "obvious".  A function (added in a separate PR?) like `modeller.getNumSolventMolecules()`, or an alternative argument `numTotal` instead of `numAdded`, would go a long way toward helping make this easy if there is the possibility that the user might start with a different number of crystallographic waters in each structure.

**Comment by peastman** [comment]
That's not really a Modeller feature.  It would make more sense on Topology.  #721 started a discussion about new Topology features, but seems to have stalled.  Perhaps we should pick up that discussion again?

**Comment by jchodera** [comment]
Sounds good to me.

**Comment by jchodera** [comment]
Though the `modeller.addSolvent(totalSolvent=X)` feature could still be a useful feature in addition to `solventAdded=X` feature you have just implemented.
